Feature/api 8858 customer sso docs

bkubiak opened the pull request:
# 🚀 Links

- [Feature branch](https://developers.labs.livechat.com/docs/feature/API-8858-customer-sso-docs)
- [Jira](https://livechatinc.atlassian.net/browse/API-8858)

# 📓 Description

We want to rewrite customer-sso (customer-accounts) documentation - open-api specification and customer authorization flow description.

ale-kacperczyk commented on content/authorization/authorizing-api-calls/index.mdx:
```diff
@@ -692,33 +692,38 @@ curl "https://accounts.livechat.com/v2/info"
 
+## Cookie grant
+
+You can receive an identity of a customer. The identity data is stored in cookies: `__lc_cid` and `__lc_cst`. These cookies are protected with `Secure` and `HttpOnly` flags so they are sent only via HTTPS and are accessible only by a specific server. They are valid for 2 years from the first request and are extended in every next call. These cookies are necessary to recognize the customer when an ...
```
```suggestion
Cookie grant is based on the `__lc_cid` and `__lc_cst` cookies, which store your customer identity data. These cookies are protected with the `Secure` and `HttpOnly` flags, which means that one can send the cookies only via HTTPS and only specific servers can access them.

The cookies are valid for 2 years, and each next call to `https://accounts.livechat.com/customer/token` extends their lifetime. The `__lc_cid` and `__lc_cst` cookies are necessary to recognize a customer when an access token is being issued.
```

ale-kacperczyk commented on content/authorization/authorizing-api-calls/index.mdx:
```diff
@@ -730,19 +735,29 @@ The response is a JSON with the following parameters:
 | `expires_in`   | A number in seconds specifying how long the `access_token` will be valid; 28800 sec by default. When it expires, you will need to generate new `access_token` using `refresh_token` (read [Using the refresh token](#using-the-refresh-token) for more details). |
 | `token_type`   | Value: `Bearer`                                                                                                                                                                                                                                                  |
 
+The response also includes important headers with cookies, for example:
```
```suggestion
The response also includes headers with the `__lc_cid` and `__lc_cst` cookies, for example:
```

ale-kacperczyk commented on content/authorization/authorizing-api-calls/index.mdx:
```diff
@@ -730,19 +735,29 @@ The response is a JSON with the following parameters:
 </Text>
 <Code>
-<CodeSample path={'CREATE A NEW CUSTOMER'}>
+<CodeSample path={'NEW CUSTOMER'}>
```
```suggestion
<CodeSample path={'COOKIE GRANT: NEW CUSTOMER'}>
```

ale-kacperczyk commented on content/authorization/authorizing-api-calls/index.mdx:
```diff
@@ -764,6 +779,126 @@ curl "https://accounts.livechat.com/customer/" \
 </Code>
 </Section>
 
+### Step 2: Existing customer
```
```suggestion
### Case: Existing customer
```

ale-kacperczyk commented on content/authorization/authorizing-api-calls/index.mdx:
```diff
@@ -764,6 +779,126 @@ curl "https://accounts.livechat.com/customer/" \
 
+### Step 2: Existing customer
+
+If you need a new access token for the identity, you have to make the same request like the one described in Step 1. However, this time, the web browser will attach cookies to the server and you'll be recognized as an existing customer. You'll get a new access token and the cookies will be extended - they will be still valid for 2 years from the last request.
```
```suggestion
Use this option to generate a new access token for an already existing customer identity.

To achieve that, make the same exact same request as in [Case: New customer](#case-new-customer-1). 

Your `entity_id`, or any other customer identity data, won' be overwritten. The web browser will attach cookies to the server allowing it to recognize you as an existing customer. You'll get a new access token, and cookies' lifetime will be extended, too. They will be valid for 2 years from your last request to `https://accounts.livechat.com/customer/token`.
```

ale-kacperczyk commented on content/authorization/authorizing-api-calls/index.mdx:
```diff
@@ -764,6 +779,126 @@ curl "https://accounts.livechat.com/customer/" \
+
+## Agent token grant
+
+You can create an identity of a new customer and later use it to perform actions on behalf of the customer. To start with this flow, you'll need an **agent access token** with a scope: `customers:own`. You can get it from one of the [Agent authorization flows](/authorization/authorizing-api-calls/#agent-authorization-flows).
```
```suggestion
Agent token grant lets you create a customer identity and use it to act on behalf of a customer. To begin, you'll need an **agent access token** with the `customers:own` scope. You can get it via one of the [Agent authorization flows](/authorization/authorizing-api-calls/#agent-authorization-flows).
```

ale-kacperczyk commented on content/authorization/authorizing-api-calls/index.mdx:
```diff
@@ -764,6 +779,126 @@ curl "https://accounts.livechat.com/customer/" \
+<Section>
+<Text>
+
+### Step 1: New customer
```
```suggestion
### Case: New customer
```

ale-kacperczyk commented on content/authorization/authorizing-api-calls/index.mdx:
```diff
@@ -764,6 +779,126 @@ curl "https://accounts.livechat.com/customer/" \
+
+### Step 1: New customer
+
+First of all, make a HTTP POST request to create a new customer identity by using the following URL:
```
```suggestion
First of all, make a HTTP POST request to create a new customer identity by calling the following URL:
```

ale-kacperczyk commented on content/authorization/authorizing-api-calls/index.mdx:
````diff
@@ -764,6 +779,126 @@ curl "https://accounts.livechat.com/customer/" \
+https://accounts.livechat.com/customer/token
+```
+
+Remember to authorize using agent access token by adding `Authorization` header to the request.
````
```suggestion
Remember to authorize your request using the `Authorization` header with an **agent access token**.
```

ale-kacperczyk commented on content/authorization/authorizing-api-calls/index.mdx:
```diff
@@ -764,6 +779,126 @@ curl "https://accounts.livechat.com/customer/" \
+
+| Parameter       | Required | Description                                                                                                                                                                                                                                                                                        |
+|-----------------|----------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| `grant_type`    | yes      | Use `agent_token` for this authorization flow                                                                                                                                                                                                                                                      |
```
```suggestion
| `grant_type`    | yes      | Use `access_token` (agent access token) for this authorization flow.                                                                                                                                                                                                                                                      |
```

ale-kacperczyk commented on content/authorization/authorizing-api-calls/index.mdx:
```diff
@@ -764,6 +779,126 @@ curl "https://accounts.livechat.com/customer/" \
+| `grant_type`    | yes      | Use `agent_token` for this authorization flow                                                                                                                                                                                                                                                      |
+| `client_id`     | yes      | Client ID of your application                                                                                                                                                                                                                                                                      |
+| `response_type` | yes      | Value: `token`                                                                                                                                                                                                                                                                                     |
+| `entity_id`     | no       | Customer ID. If you don't specify it, a new customer will be created (Step 1). Specify it (described in Step 2) if you want to [acquire a new token for an existing customer](/getting-started/authorization/faq/#the-access-token-of-my-customer-has-expired-do-i-need-to-create-a-new-customer). |
```
```suggestion
| `entity_id`     | no       | Customer ID. If you don't specify it, a new customer identity will be created (Case: New customer). Specify it if you want to [acquire a new token for an existing customer identity](/getting-started/authorization/faq/#the-access-token-of-my-customer-has-expired-do-i-need-to-create-a-new-customer). |
```

bkubiak replied:
I also added a link to the case

ale-kacperczyk commented on content/authorization/authorizing-api-calls/index.mdx:
```diff
@@ -764,6 +779,126 @@ curl "https://accounts.livechat.com/customer/" \
+| `access_token` | A token you can use to call LiveChat APIs on behalf of the user (customer).                                                                                                                                                                                      |
+| `client_id`    | **Client Id** of your app                                                                                                                                                                                                                                        |
+| `entity_id`    | The ID of the newly created customer                                                                                                                                                                                                                             |
+| `expires_in`   | A number in seconds specifying how long the `access_token` will be valid; 28800 sec by default. When it expires, you will need to generate new `access_token` using `refresh_token` (read [Using the refresh token](#using-the-refresh-token) for more details). |
```
```suggestion
| `expires_in`   | A number in seconds specifying how long your customer `access_token` will be valid; 28800 sec by default. When it expires, you will need to generate new `access_token` using `refresh_token` (read [Using the refresh token](#using-the-refresh-token) for more details). |
```

bkubiak replied:
There was a little mistake here, I changed it to:

A number in seconds specifying how long the `access_token` will be valid; 28800 sec by default. When it expires, you will need to generate new `access_token` following [Case: Existing customer](#case-existing-customer-1)).

ale-kacperczyk commented on content/authorization/authorizing-api-calls/index.mdx:
```diff
@@ -764,6 +779,126 @@ curl "https://accounts.livechat.com/customer/" \
+
+</Text>
+<Code>
+<CodeSample path={'NEW CUSTOMER'}>
```
```suggestion
<CodeSample path={'AGENT TOKEN GRANT: NEW CUSTOMER'}>
```

ale-kacperczyk commented on content/authorization/authorizing-api-calls/index.mdx:
```diff
@@ -764,6 +779,126 @@ curl "https://accounts.livechat.com/customer/" \
+<Section>
+<Text>
+
+### Step 2: Existing customer
```
```suggestion
### Case: Existing customer
```

ale-kacperczyk commented on content/authorization/authorizing-api-calls/index.mdx:
```diff
@@ -764,6 +779,126 @@ curl "https://accounts.livechat.com/customer/" \
+
+### Step 2: Existing customer
+
+After creating a new customer identity and using it to perform actions on behalf of the customer, you may want to acquire a new access token for the existing customer - the one you have just created. Make the same request as earlier but this time add an additional parameter: `entity_id`. It can be found in the response in Step 1.
```
```suggestion
Use this case to get a new access token for an already existing customer identity. 

Make the same request as in [Case: New customer](#case-new-customer-2), but this time, specify `entity_id` to avoid creating a new identity. You can find your `entity_id` [in the response](#response-5) of the request to create a new customer identity.
```

bkubiak replied:
I guess the link should be `#case-new-customer-1`

ale-kacperczyk replied:
I'm afraid #case-new-customer-1 will link to Cookie grant https://developers.labs.livechat.com/docs/feature/API-8858-customer-sso-docs/authorization/authorizing-api-calls/#step-1-new-customer

ale-kacperczyk commented on content/authorization/authorizing-api-calls/index.mdx:
```diff
@@ -764,6 +779,126 @@ curl "https://accounts.livechat.com/customer/" \
+
+</Text>
+<Code>
+<CodeSample path={'EXISTING CUSTOMER'}>
```
```suggestion
<CodeSample path={'AGENT TOKEN GRANT: EXISTING CUSTOMER'}>
```

ale-kacperczyk reviewed (commented):
I've only checked the descriptive part. Open API is still to do for me.

ale-kacperczyk commented on src/configs/redoc/livechat-customer-accounts-api/spec.yml:
```diff
@@ -0,0 +1,249 @@
+openapi: 3.0.0
+info:
+  title: Customer Accounts API
```
One general remark is that I think that the file name should also be "Customer Accounts API", without "LiveChat"

ale-kacperczyk replied:
I'll rename it

bkubiak commented:
I have fixed all copy issues.

ale-kacperczyk commented on content/authorization/authorizing-api-calls/index.mdx:
```diff
@@ -692,33 +692,44 @@ curl "https://accounts.livechat.com/v2/info"
 To authorize requests to the [Customer Chat API](/messaging/customer-chat-api/), you'll need a different access token than the one used for the [Agent Chat API](/messaging/agent-chat-api/) and [Configuration API](/management/configuration-api/).
 Throughout the whole API documentation, we'll refer to it as `customer_access_token`.
 
+Customer authorization flows allow you to receive an identity of a customer. You can do it via [Cookie grant](#cookie-grant) if you want to receive an identity of a single customer or [Agent token grant](#agent-token-grant) in order to manage indentities of multiple customers.
```
```suggestion
Customer authorization flows allow you to receive an identity of a customer. You can do it via [Cookie grant](#cookie-grant) if you want to receive an identity of a single customer or via [Agent token grant](#agent-token-grant) in order to manage multiple customer identities.
```

ale-kacperczyk commented on content/authorization/authorizing-api-calls/index.mdx:
```diff
@@ -692,33 +692,44 @@ curl "https://accounts.livechat.com/v2/info"
+
+The cookies are valid for 2 years, and each next call to `https://accounts.livechat.com/customer/token` extends their lifetime. The `__lc_cid` and `__lc_cst` cookies are necessary to recognize a customer when an access token is being issued.
+
+This flow is recommended for front-end apps that support cookies (like an app running in a web browser). Each user of the app receives an identity if a single customer.
```
```suggestion
This flow is recommended for frontend apps that support cookies, such as an app running in a web browser. It has its limitations – all app users will share the same customer identity (the one whose data is stored in cookies).
```

ale-kacperczyk replied:
I'm not sure if I got the meaning right

bkubiak replied:
Not really - I meant that every client (user) of the frontend app will get its own, unique identity and they will be able to issue an access token using this identity. So there will be many customer identities but each of them will be managed separately by its owner

ale-kacperczyk replied:
I understood something different from our conversation earlier 😅 That Cookie grant limits customer identities to 1 because the browser will not store multiple customers' data at the same time.

PS: What do you mean by owner?

bkubiak replied:
In the agent-token flow, owner is the backend integration and it owns many identities. In the cookie flow, owner of the identity is a user of the frontend app, however, there are many users.

Anyway, I have tweaked it a bit. Does it look better, now?

ale-kacperczyk replied:
Okay, I think I get it finally :D I'll get rid of that info in brackets because it adds more confusion `(the one whose data is stored in cookies)` than value

ale-kacperczyk replied:
I'll also add one short sentence to Agent token auth that this flow is recommended for backend apps, so that it's straightforward

ale-kacperczyk commented on content/authorization/authorizing-api-calls/index.mdx:
```diff
@@ -764,6 +785,140 @@ curl "https://accounts.livechat.com/customer/" \
+
+Agent token grant lets you create a customer identity and use it to act on behalf of a customer. To begin, you'll need an **agent access token** with the `customers:own` scope. You can get it via one of the [Agent authorization flows](/authorization/authorizing-api-calls/#agent-authorization-flows).
+
+We recommend this flow if you want to to impersonate customers and simulate them sending messages from other systems, like Facebook Messenger or WhatsApp, in LiveChat. An integration that follows this authorization flow behaves like a middleman between these systems and LiveChat and manages identities of multiple customers.
```
```suggestion
We recommend this flow if you want to to impersonate customers and simulate them sending messages from external systems, like Facebook Messenger or WhatsApp, in LiveChat. An integration that follows this authorization flow acts as s middleman between an external system and LiveChat, and it has the ability to manage multiple customer identities.
```

ale-kacperczyk commented on content/authorization/authorizing-api-calls/index.mdx:
```diff
@@ -764,6 +785,140 @@ curl "https://accounts.livechat.com/customer/" \
+
+## Revoking tokens
+
+Look at the [Customer Accounts API specification](/authorization/customer-accounts-api/#tag/Tokens/paths/~1token/delete) to find more details about revoking tokens.
```
```suggestion
Look at the [Customer Accounts API specification](/authorization/customer-accounts-api/#tag/Tokens/paths/~1token/delete) to find more details about revoking customer tokens.
```

ale-kacperczyk commented on content/authorization/authorizing-api-calls/index.mdx:
```diff
@@ -764,6 +785,140 @@ curl "https://accounts.livechat.com/customer/" \
+
+## Validating the access token
+
+The instructions on how to validate access token of a customer are available in [Customer Accounts API specification](/authorization/customer-accounts-api/#tag/Tokens/paths/~1info/get).
```
```suggestion
The instruction on how to validate customer access token of a customer is available in the [Customer Accounts API specification](/authorization/customer-accounts-api/#tag/Tokens/paths/~1info/get).
```

bkubiak commented:
I fixed all copy issues, only this one is waiting for a suggestion: https://github.com/livechat/livechat-public-docs/pull/964#discussion_r656292630

ale-kacperczyk commented on src/configs/redoc/customer-accounts-api/spec.yml:
```diff
@@ -0,0 +1,249 @@
+openapi: 3.0.0
+info:
+  title: Customer Accounts API
+  description: API responsible for authorization and authentication processes of customers.
```
```suggestion
  description: Customer Accounts API is responsible for the authentication and authorization processes of customers in LiveChat and HelpDesk.
```

bkubiak replied:
just LiveChat

ale-kacperczyk commented on src/configs/redoc/customer-accounts-api/spec.yml:
```diff
@@ -0,0 +1,249 @@
+        authorizationCode:
+          authorizationUrl: 'https://accounts.livechat.com'
+          tokenUrl: 'https://accounts.livechat.com/token'
+    Cookie Auth:
```
```suggestion
    Cookie Grant:
```

bkubiak replied:
I have changed the order but I prefer "Auth" here, this will still be true, take a look please: https://swagger.io/docs/specification/authentication/cookie-authentication/

ale-kacperczyk replied:
So maybe we should name our this flow in our docs "Cookie Auth"? Or "Cookie Auth Grant"?

ale-kacperczyk commented on src/configs/redoc/customer-accounts-api/spec.yml:
```diff
@@ -0,0 +1,249 @@
+          authorizationUrl: 'https://accounts.livechat.com/customer'
+    OAuth2 Agent Bearer Token:
+      description: >-
+        This API uses OAuth2 for agents with the implicit grant flow. <a
```
```suggestion
        This API uses OAuth2 for agents with the Implicit grant flow. <a
```

ale-kacperczyk commented on src/configs/redoc/customer-accounts-api/spec.yml:
```diff
@@ -0,0 +1,249 @@
+      description: >-
+        This API uses OAuth2 for agents with the implicit grant flow. <a
+        href="https://developers.livechat.com/docs/authorization/authorizing-api-calls/#implicit-grant">Learn
+        about the implicit grant flow.</a>
```
```suggestion
        about the Implicit grant flow.</a>
```

ale-kacperczyk commented on src/configs/redoc/customer-accounts-api/spec.yml:
```diff
@@ -0,0 +1,249 @@
+  /info:
+    get:
+      summary: Get token details
+      description: The access token can be provided via authorization or a query string.
```
```suggestion
      description: Retrieve information about an access token. You can provide the token via authorization or a query string.
```

ale-kacperczyk commented on src/configs/redoc/customer-accounts-api/spec.yml:
```diff
@@ -0,0 +1,249 @@
+  /token:
+    post:
+      summary: Create a token
+      description: Creates a token
```
```suggestion
      description: Creates a new customer access token.
```

ale-kacperczyk commented on src/configs/redoc/customer-accounts-api/spec.yml:
```diff
@@ -0,0 +1,249 @@
+      summary: Revoke token
+      description: >-
+        Revoke a token provided via authorization
+        The access token can be provided via authorization.
```
```suggestion
```

ale-kacperczyk commented on src/configs/redoc/customer-accounts-api/spec.yml:
```diff
@@ -0,0 +1,249 @@
+  schemas: {}
+  securitySchemes:
+    OAuth2 Bearer Token:
+      description: This API uses OAuth2 token for customers
```
```suggestion
      description: This API uses OAuth2 tokens.
```

ale-kacperczyk reviewed (commented):
Changes for the spec itself

bkubiak commented:
I have fixed all open-api issues. And updated a bit description of cookie grant (https://github.com/livechat/livechat-public-docs/pull/964#discussion_r656593943).

jakubsikora reviewed (approved):
Nice!
